src/native/libs/System.Native/pal_networking.c

```diff
@@ -2762,6 +2765,97 @@ int32_t SystemNative_GetBytesAvailable(intptr_t socket, int32_t* available)
     return Error_SUCCESS;
 }
 
+int32_t SystemNative_Select(int* readFds, int readFdsCount, int* writeFds, int writeFdsCount,  int* errorFds, int errorFdsCount, int32_t microseconds, int maxFd, int* triggered)
+{
+#ifdef _DARWIN_UNLIMITED_SELECT
+    fd_set readSet;
+    fd_set writeSet;
+    fd_set errorSet;
+
+    fd_set* readSetPtr;
+    fd_set* writeSetPtr;
+    fd_set* errorSetPtr;
+
+    if (maxFd < FD_SETSIZE)
+    {
+        FD_ZERO(&readSet);
+        FD_ZERO(&writeSet);
+        FD_ZERO(&errorSet);
+        readSetPtr = readFdsCount == 0 ? NULL : &readSet;
+        writeSetPtr= writeFdsCount == 0 ? NULL : &writeSet;
+        errorSetPtr = errorFdsCount == 0 ? NULL : &errorSet;
+    }
+    else
+    {
+       readSetPtr = readFdsCount == 0 ? NULL : calloc( __DARWIN_howmany(maxFd, __DARWIN_NFDBITS),  sizeof(int32_t));
+       writeSetPtr = writeFdsCount == 0 ? NULL : calloc( __DARWIN_howmany(maxFd, __DARWIN_NFDBITS),  sizeof(int32_t));
+       errorSetPtr = errorFdsCount == 0 ? NULL :calloc( __DARWIN_howmany(maxFd, __DARWIN_NFDBITS),  sizeof(int32_t));
+    }
+
+
+    struct timeval timeout;
+    timeout.tv_sec = microseconds / 1000000;
+    timeout.tv_usec = microseconds % 1000000;
+
+    int fd;
+    for (int i = 0 ; i < readFdsCount; i++)
+    {
+        fd = *(readFds + i);
+        __DARWIN_FD_SET(fd, readSetPtr);
+    }
+    for (int i = 0 ; i < writeFdsCount; i++)
+    {
+        fd = *(writeFds + i);
+        __DARWIN_FD_SET(fd, writeSetPtr);
+    }
+    for (int i = 0 ; i < errorFdsCount; i++)
+    {
+        fd = *(errorFds + i);
+        __DARWIN_FD_SET(fd, errorSetPtr);
+    }
+
+    *triggered = select(maxFd + 1, readSetPtr, writeSetPtr, errorSetPtr, microseconds < 0 ? NULL : &timeout);
+
+    if (*triggered < 0)
+    {
+        return SystemNative_ConvertErrorPlatformToPal(errno);
+    }
+
+    for (int i = 0 ; i < readFdsCount; i++)
+    {
+        fd = *(readFds + i);
+        *(readFds + i) = __DARWIN_FD_ISSET(fd, readSetPtr);
```

Comment:
What does __DARWIN_FD_ISSET return? Is it returning fd if it's set and otherwise 0?

Reply:
this is basically same as `FD_ISSET` but working on set larger than FD_SETSIZE. man page says 

> FD_ISSET(fd, &fdset) is non-zero if fd is a member of fdset, zero otherwise.

in practice this seems to be int with one bit set depending on the position in it bit array.